File: backend/app/schemas/submission.py
from pydantic import BaseModel, Field, field_serializer
from datetime import datetime, timezone
from typing import Optional, List, Dict, Any
from app.models.submission import SessionStatus, SubmissionStatus


class ExamSessionBase(BaseModel):
    """Base exam session schema"""
    exam_id: str
    student_id: str


class ExamSessionCreate(ExamSessionBase):
    """Schema for creating exam session"""
    pass


class ExamSession(ExamSessionBase):
    """Exam session response schema"""
    id: str
    variant_seed: int
    variant_assignments: Dict[str, str]
    started_at: datetime
    submitted_at: Optional[datetime] = None
    expires_at: datetime
    status: SessionStatus
    attempt_number: int
    
    # Serialize datetime fields as UTC with 'Z' suffix
    @field_serializer('started_at', 'submitted_at', 'expires_at')
    def serialize_datetime(self, dt: Optional[datetime], _info) -> Optional[str]:
        if dt is None:
            return None
        # If naive datetime, assume it's UTC
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt.isoformat()
    
    class Config:
        from_attributes = True


class SubmissionImageBase(BaseModel):
    """Base submission image schema"""
    filename: str
    order_index: int


class SubmissionImage(SubmissionImageBase):
    """Submission image response schema"""
    id: str
    file_path: str
    file_size: int
    mime_type: str
    is_processed: bool
    quality_score: Optional[float] = None
    uploaded_at: datetime
    
    # Serialize datetime fields as UTC with 'Z' suffix
    @field_serializer('uploaded_at')
    def serialize_datetime(self, dt: Optional[datetime], _info) -> Optional[str]:
        if dt is None:
            return None
        # If naive datetime, assume it's UTC
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt.isoformat()
    
    class Config:
        from_attributes = True


class SubmissionScoreBase(BaseModel):
    """Base submission score schema"""
    task_type_id: str
    criterion_name: str
    criterion_description: Optional[str] = None
    max_score: float


class SubmissionScore(SubmissionScoreBase):
    """Submission score response schema"""
    id: str
    submission_id: str
    ai_score: Optional[float] = None
    final_score: Optional[float] = None
    ai_comment: Optional[str] = None
    teacher_comment: Optional[str] = None
    
    class Config:
        from_attributes = True


class SubmissionBase(BaseModel):
    """Base submission schema"""
    session_id: str


class SubmissionCreate(SubmissionBase):
    """Schema for creating submission"""
    pass


class Submission(SubmissionBase):
    """Submission response schema"""
    id: str
    student_id: str
    submitted_at: datetime
    status: SubmissionStatus
    ai_score: Optional[float] = None
    final_score: Optional[float] = None
    max_score: float
    ai_analysis: Optional[Dict[str, Any]] = None
    ai_comments: Optional[str] = None
    teacher_comments: Optional[str] = None
    is_flagged: bool
    flag_reasons: List[str] = []
    reviewed_by: Optional[str] = None
    reviewed_at: Optional[datetime] = None
    images: List[SubmissionImage] = []
    scores: List[SubmissionScore] = []
    
    # Serialize datetime fields as UTC with 'Z' suffix
    @field_serializer('submitted_at', 'reviewed_at')
    def serialize_datetime(self, dt: Optional[datetime], _info) -> Optional[str]:
        if dt is None:
            return None
        # If naive datetime, assume it's UTC
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=timezone.utc)
        return dt.isoformat()
    
    class Config:
        from_attributes = True


class SubmissionUpdate(BaseModel):
    """Schema for updating submission"""
    final_score: Optional[float] = None
    teacher_comments: Optional[str] = None
    status: Optional[SubmissionStatus] = None


class SubmissionApprove(BaseModel):
    """Schema for approving submission"""
    teacher_comments: Optional[str] = None


class SubmissionOverride(BaseModel):
    """Schema for overriding submission score"""
    final_score: float
    teacher_comments: str
    scores: Optional[List[Dict[str, Any]]] = None

File: src/schemas/submission.rs
use serde::{Deserialize, Serialize};
use validator::Validate;

pub(crate) use crate::core::time::format_primitive;
use crate::db::types::{
    LlmPrecheckStatus, OcrImageStatus, OcrIssueSeverity, OcrOverallStatus, OcrPageStatus,
    SessionStatus, SubmissionStatus,
};

#[derive(Debug, Serialize)]
pub(crate) struct ExamSessionResponse {
    pub(crate) id: String,
    pub(crate) course_id: String,
    pub(crate) exam_id: String,
    pub(crate) student_id: String,
    pub(crate) variant_seed: i32,
    pub(crate) variant_assignments: serde_json::Value,
    pub(crate) started_at: String,
    pub(crate) submitted_at: Option<String>,
    pub(crate) expires_at: String,
    pub(crate) status: SessionStatus,
    pub(crate) attempt_number: i32,
}

#[derive(Debug, Serialize)]
pub(crate) struct SubmissionImageResponse {
    pub(crate) id: String,
    pub(crate) course_id: String,
    pub(crate) filename: String,
    pub(crate) order_index: i32,
    pub(crate) file_size: i64,
    pub(crate) mime_type: String,
    pub(crate) is_processed: bool,
    pub(crate) ocr_status: OcrImageStatus,
    pub(crate) ocr_text: Option<String>,
    pub(crate) ocr_markdown: Option<String>,
    pub(crate) ocr_chunks: Option<serde_json::Value>,
    pub(crate) quality_score: Option<f64>,
    pub(crate) uploaded_at: String,
}

#[derive(Debug, Serialize)]
pub(crate) struct SubmissionScoreResponse {
    pub(crate) id: String,
    pub(crate) course_id: String,
    pub(crate) submission_id: String,
    pub(crate) task_type_id: String,
    pub(crate) criterion_name: String,
    pub(crate) criterion_description: Option<String>,
    pub(crate) ai_score: Option<f64>,
    pub(crate) final_score: Option<f64>,
    pub(crate) ai_comment: Option<String>,
    pub(crate) teacher_comment: Option<String>,
    pub(crate) max_score: f64,
}

#[derive(Debug, Serialize)]
pub(crate) struct SubmissionResponse {
    pub(crate) id: String,
    pub(crate) course_id: String,
    pub(crate) session_id: String,
    pub(crate) student_id: String,
    pub(crate) submitted_at: String,
    pub(crate) status: SubmissionStatus,
    pub(crate) ocr_overall_status: OcrOverallStatus,
    pub(crate) llm_precheck_status: LlmPrecheckStatus,
    pub(crate) report_flag: bool,
    pub(crate) report_summary: Option<String>,
    pub(crate) ocr_error: Option<String>,
    pub(crate) llm_error: Option<String>,
    pub(crate) ai_score: Option<f64>,
    pub(crate) final_score: Option<f64>,
    pub(crate) max_score: f64,
    pub(crate) ai_analysis: Option<serde_json::Value>,
    pub(crate) ai_comments: Option<String>,
    pub(crate) teacher_comments: Option<String>,
    pub(crate) is_flagged: bool,
    pub(crate) flag_reasons: Vec<String>,
    pub(crate) reviewed_by: Option<String>,
    pub(crate) reviewed_at: Option<String>,
    pub(crate) next_step: Option<SubmissionNextStep>,
    pub(crate) images: Vec<SubmissionImageResponse>,
    pub(crate) scores: Vec<SubmissionScoreResponse>,
}

#[derive(Debug, Clone, Copy, Serialize)]
#[serde(rename_all = "snake_case")]
pub(crate) enum SubmissionNextStep {
    OcrReview,
    Result,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub(crate) struct OcrIssueDto {
    pub(crate) anchor: serde_json::Value,
    #[serde(default)]
    pub(crate) original_text: Option<String>,
    #[serde(default)]
    pub(crate) suggested_text: Option<String>,
    pub(crate) note: String,
    #[serde(default = "default_ocr_issue_severity")]
    pub(crate) severity: OcrIssueSeverity,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct OcrIssueResponse {
    pub(crate) id: String,
    pub(crate) review_id: String,
    pub(crate) image_id: String,
    pub(crate) anchor: serde_json::Value,
    pub(crate) original_text: Option<String>,
    pub(crate) suggested_text: Option<String>,
    pub(crate) note: String,
    pub(crate) severity: OcrIssueSeverity,
    pub(crate) created_at: String,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct OcrPageResponse {
    pub(crate) image_id: String,
    pub(crate) image_view_url: Option<String>,
    pub(crate) ocr_status: OcrImageStatus,
    pub(crate) ocr_markdown: Option<String>,
    pub(crate) chunks: Option<serde_json::Value>,
    pub(crate) page_status: Option<OcrPageStatus>,
    pub(crate) issues: Vec<OcrIssueResponse>,
}

#[derive(Debug, Clone, Serialize)]
pub(crate) struct OcrPagesResponse {
    pub(crate) submission_id: String,
    pub(crate) ocr_status: OcrOverallStatus,
    pub(crate) llm_precheck_status: LlmPrecheckStatus,
    pub(crate) report_flag: bool,
    pub(crate) report_summary: Option<String>,
    pub(crate) pages: Vec<OcrPageResponse>,
}

#[derive(Debug, Deserialize, Validate)]
pub(crate) struct OcrReviewUpsertRequest {
    pub(crate) page_status: OcrPageStatus,
    #[serde(default)]
    pub(crate) issues: Vec<OcrIssueDto>,
}

#[derive(Debug, Clone, Copy, Deserialize, Serialize, PartialEq, Eq)]
#[serde(rename_all = "lowercase")]
pub(crate) enum OcrFinalizeAction {
    Submit,
    Report,
}

#[derive(Debug, Deserialize, Validate)]
pub(crate) struct FinalizeOcrReviewRequest {
    pub(crate) action: OcrFinalizeAction,
    #[serde(default)]
    pub(crate) report_summary: Option<String>,
}

#[derive(Debug, Deserialize)]
pub(crate) struct SubmissionApproveRequest {
    #[serde(default)]
    pub(crate) teacher_comments: Option<String>,
}

#[derive(Debug, Deserialize, Validate)]
pub(crate) struct SubmissionOverrideRequest {
    #[validate(range(min = 0.0, message = "final_score must be non-negative"))]
    pub(crate) final_score: f64,
    pub(crate) teacher_comments: String,
}

fn default_ocr_issue_severity() -> OcrIssueSeverity {
    OcrIssueSeverity::Major
}
